druid-shell/src/platform/web/window.rs

```diff
@@ -0,0 +1,633 @@
+        let delta_mode = event.delta_mode();
+
+        let dx = event.delta_x();
+        let dy = event.delta_y();
+        let height = state.canvas.height() as f64;
+        let width = state.canvas.width() as f64;
+
+        let modifiers = get_modifiers!(event);
+        let mut handler = state.handler.borrow_mut();
+
+        // The value 35.0 was manually picked to produce similar behavior to mac/linux.
+        match delta_mode {
+            web_sys::WheelEvent::DOM_DELTA_PIXEL => handler.wheel(Vec2::from((dx, dy)), modifiers),
+            web_sys::WheelEvent::DOM_DELTA_LINE => {
+                handler.wheel(Vec2::from((35.0 * dx, 35.0 * dy)), modifiers)
+            }
+            web_sys::WheelEvent::DOM_DELTA_PAGE => {
+                handler.wheel(Vec2::from((width * dx, height * dy)), modifiers)
+            }
+            _ => warn!("Invalid deltaMode in WheelEvent: {}", delta_mode),
+        }
+    });
+}
+
+fn setup_resize_callback(ws: &Rc<WindowState>) {
+    let state = ws.clone();
+    register_window_event_listener(ws, "resize", move |_: web_sys::UiEvent| {
+        let (css_width, css_height, dpr) = state.get_window_size_and_dpr();
+        let physical_width = (dpr * css_width) as u32;
+        let physical_height = (dpr * css_height) as u32;
+        state.dpr.replace(dpr);
+        state.canvas.set_width(physical_width);
+        state.canvas.set_height(physical_height);
+        let _ = state.context.scale(dpr, dpr);
+        state
+            .handler
+            .borrow_mut()
+            .size(physical_width, physical_height);
+    });
+}
+
+fn setup_keyup_callback(ws: &Rc<WindowState>) {
+    let state = ws.clone();
+    register_window_event_listener(ws, "keyup", move |event: web_sys::KeyboardEvent| {
+        let code = KeyCode::from((event.key_code(), event.location()));
+        let mods = get_modifiers!(event);
+        let text = key_to_text(event.key());
+        let repeat = event.repeat();
+        let event = keyboard::KeyEvent::new(code, repeat, mods, text.as_str(), text.as_str());
+        state.handler.borrow_mut().key_up(event);
+    });
+}
+
+fn setup_keydown_callback(ws: &Rc<WindowState>) {
+    let state = ws.clone();
+    register_window_event_listener(ws, "keydown", move |event: web_sys::KeyboardEvent| {
+        let code = KeyCode::from((event.key_code(), event.location()));
+        let mods = get_modifiers!(event);
+        let text = key_to_text(event.key());
+        let repeat = event.repeat();
+        if let KeyCode::Backspace = code {
+            // Prevent the browser from going back a page by default.
+            event.prevent_default();
+        }
+        let event = keyboard::KeyEvent::new(code, repeat, mods, text.as_str(), text.as_str());
+        state.handler.borrow_mut().key_down(event);
+    });
+}
+
+/// A helper to convert the key string to the text it is supposed to represent.
+fn key_to_text(key: String) -> String {
```

Comment:
total code golf but I'd do this as,

```rust
fn key_to_text(key: &str) -> &str {
    match key {
    k if k.len() == 1 => k,
    "Enter" => "\n",
    _ => "",
    }
}
```

Also: what about tab?

Also: this length check suppresses all non-ascii input. For instance I can't type é or ô in the hello example. (thanks for providing the examples!)

Reply:
Good find! I was afraid of limiting the output lifetime of the string, but it seems like this isn't problematic for this case.

As for the last two points, we can match through all of https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key/Key_Values and adjust the key "text" appropriately.